src/_pytest/tmpdir.py

```diff
@@ -33,46 +128,13 @@ def mktemp(self, basename, numbered=True):
         If ``numbered``, ensure the directory is unique by adding a number
         prefix greater than any existing one.
         """
-        basetemp = self.getbasetemp()
-        if not numbered:
-            p = basetemp.mkdir(basename)
-        else:
-            p = py.path.local.make_numbered_dir(
-                prefix=basename, keep=0, rootdir=basetemp, lock_timeout=None
-            )
-        self.trace("mktemp", p)
-        return p
+        return py.path.local(self.tmppath_factory.mktemp(basename, numbered).resolve())
 
     def getbasetemp(self):
-        """ return base temporary directory. """
-        try:
-            return self._basetemp
-        except AttributeError:
-            basetemp = self.config.option.basetemp
-            if basetemp:
-                basetemp = py.path.local(basetemp)
-                if basetemp.check():
-                    basetemp.remove()
-                basetemp.mkdir()
-            else:
-                temproot = py.path.local.get_temproot()
-                user = get_user()
-                if user:
-                    # use a sub-directory in the temproot to speed-up
-                    # make_numbered_dir() call
-                    rootdir = temproot.join("pytest-of-%s" % user)
-                else:
-                    rootdir = temproot
-                rootdir.ensure(dir=1)
-                basetemp = py.path.local.make_numbered_dir(
-                    prefix="pytest-", rootdir=rootdir
-                )
-            self._basetemp = t = basetemp.realpath()
-            self.trace("new basetemp", t)
-            return t
+        return py.path.local(self.tmppath_factory.getbasetemp().resolve())
```

Comment:
ah ok, so it still returns a `py.path` for now?

Reply:
the shim and the old api must stay compatible

Reply:
makes sense, another approach might be to write an adapter class which has the interface of `py.path._Path` but keeps a `pathlib` object internally -- I suspect we'll want something like this anyway to make a path forward for methods like `as_cwd()` which have no equivalent in `pathlib`

Reply:
I'd be a little sad to see some of the apis go.  I for one use:

```python
with tmpdir.join('somepath').ensure().as_cwd():
    ...
```

quite a lot

Reply:
ensure is going to go away as well - i wonder where you use that patter, since its a bit pathological in common usage and i consider it a anti-pattern

Reply:
most of my tests, actually 😆 

```
$ all-repos-grep -l -E '\.(as_cwd|ensure(|_dir))\(\)'
repos/asottile/all-repos:tests/push/github_pull_request_test.py
repos/asottile/all-repos-depends:tests/conftest.py
repos/asottile/all-repos-depends:tests/depends_test.py
repos/asottile/all-repos-depends:tests/lang/python_test.py
repos/asottile/all-repos-depends:tests/packages_test.py
repos/asottile/aspy.refactor_imports:tests/classify_test.py
repos/asottile/aspy.refactor_imports:tests/conftest.py
repos/asottile/aspy.refactor_imports:tests/sort_test.py
repos/asottile/git-code-debt:tests/server/app_test.py
repos/asottile/no-manylinux1:tests/integration_test.py
repos/asottile/reorder_python_imports:tests/reorder_python_imports_test.py
repos/asottile/rubyvenv:tests/rubyvenv_test.py
repos/asottile/seed-isort-config:tests/seed_isort_config_test.py
repos/asottile/yesqa:tests/yesqa_test.py
repos/pre-commit/pre-commit:tests/commands/run_test.py
repos/pre-commit/pre-commit:tests/git_test.py
repos/pre-commit/pre-commit:tests/languages/pygrep_test.py
repos/pre-commit/pre-commit:tests/parse_shebang_test.py
repos/pre-commit/pre-commit:tests/repository_test.py
repos/pre-commit/pre-commit:tests/staged_files_only_test.py
repos/pre-commit/pre-commit-hooks:tests/autopep8_wrapper_test.py
repos/pre-commit/pre-commit-hooks:tests/check_added_large_files_test.py
repos/pre-commit/pre-commit-hooks:tests/check_case_conflict_test.py
repos/pre-commit/pre-commit-hooks:tests/check_merge_conflict_test.py
repos/pre-commit/pre-commit-hooks:tests/check_symlinks_test.py
repos/pre-commit/pre-commit-hooks:tests/check_vcs_permalinks_test.py
repos/pre-commit/pre-commit-hooks:tests/forbid_new_submodules_test.py
repos/pre-commit/pre-commit-hooks:tests/no_commit_to_branch_test.py
repos/pre-commit/pre-commit-mirror-maker:tests/make_repo_test.py
repos/sass/libsass-python:sasstests.py
```

Reply:
@asottile should i fix them?

Reply:
ehhhhhh I can fix them if the apis go away